File: backend/monitor/contracts.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Literal, Protocol
from pydantic import BaseModel, ConfigDict, Field, model_validator


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


class NormalizedEvent(BaseModel):
    model_config = ConfigDict(extra="forbid")
    source_id: str
    provider_record_id: str
    kind: Literal["incident", "advisory", "vulnerability_notice", "measurement"]
    category: Literal["earthquake", "disaster", "weather", "aviation", "cyber", "internet", "space_weather"]
    title: str = Field(min_length=1, max_length=800)
    description: str = Field(default="", max_length=12000)
    source_url: str
    occurred_start: datetime | None = None
    occurred_end: datetime | None = None
    issued_at: datetime | None = None
    source_updated_at: datetime | None = None
    valid_from: datetime | None = None
    valid_to: datetime | None = None
    countries: list[str] = Field(default_factory=list)
    geometry: dict[str, Any] | None = None
    location_precision: Literal["point", "area", "country", "unknown"] = "unknown"
    time_precision: Literal["second", "minute", "hour", "day", "unknown"] = "unknown"
    severity: int = Field(default=0, ge=0, le=4)
    original_severity: str | None = None
    severity_reason: str = ""
    lifecycle_status: Literal["active", "expired", "withdrawn", "unknown"] = "active"
    verification_status: str = "reported"
    origins: list[str] = Field(default_factory=list)
    external_ids: list[str] = Field(default_factory=list)
    supersedes: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    raw: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def timestamps_have_offsets(self):
        for field in ("occurred_start", "occurred_end", "issued_at", "source_updated_at", "valid_from", "valid_to"):
            value = getattr(self, field)
            if value is not None and value.tzinfo is None:
                raise ValueError(f"{field} requires timezone")
        if self.geometry is None and self.location_precision == "point":
            self.location_precision = "unknown"
        self.countries = sorted(set(c.upper() for c in self.countries))
        return self


class ProviderBatch(BaseModel):
    events: list[NormalizedEvent]
    warnings: list[str] = Field(default_factory=list)
    rejected_count: int = 0
    metadata: dict[str, Any] = Field(default_factory=dict)


class SourceSpec(BaseModel):
    id: str
    name: str
    poll_interval_seconds: int
    coverage: str
    license_name: str
    license_url: str
    attribution: str
    requires_key: bool = False


@dataclass
class FetchedDocument:
    body: bytes
    content_type: str
    url: str
    status: int = 200
    fetched_at: datetime | None = None
    not_modified: bool = False


class Fetcher(Protocol):
    async def get(self, url: str, headers: dict[str, str] | None = None) -> FetchedDocument: ...


class EventQuery(BaseModel):
    model_config = ConfigDict(extra="forbid")
    window_hours: int = Field(default=24, ge=1, le=720)
    time_basis: Literal["occurred", "changed", "published", "validity"] = "occurred"
    since: datetime | None = None
    until: datetime | None = None
    country: str | None = Field(default=None, pattern=r"^[A-Z]{2}$")
    region: Literal["europe"] | None = None
    category: Literal["earthquake", "disaster", "weather", "aviation", "cyber", "internet", "space_weather"] | None = None
    severity_min: int = Field(default=0, ge=0, le=4)
    min_sources: int = Field(default=1, ge=1, le=10)
    lat: float | None = Field(default=None, ge=-90, le=90)
    lon: float | None = Field(default=None, ge=-180, le=180)
    radius_km: float | None = Field(default=None, gt=0, le=20000)
    include_inactive: bool = False
    limit: int = Field(default=300, ge=1, le=1000)

    @model_validator(mode="after")
    def coherent_filters(self):
        radius = (self.lat, self.lon, self.radius_km)
        if any(v is not None for v in radius) and not all(v is not None for v in radius):
            raise ValueError("lat, lon and radius_km are required together")
        for value in (self.since, self.until):
            if value is not None and value.tzinfo is None:
                raise ValueError("Query timestamps require timezone")
        if self.since and self.until and self.since >= self.until:
            raise ValueError("since must precede until")
        return self


class QueryInterpretation(BaseModel):
    supported: bool
    query: EventQuery | None = None
    explanation: str
    limitations: list[str] = Field(default_factory=list)

File: web/lib/contracts.ts
import type { Geometry } from "geojson";

export type Category = "earthquake" | "disaster" | "weather" | "aviation" | "cyber" | "internet";
export type TimeBasis = "occurred" | "changed" | "published" | "validity";
export type SourceState = "pending" | "ok" | "ok_empty" | "partial" | "error" | "stale" | "needs_credentials" | "disabled";
export interface SourceStatus {
  id: string; name: string; status: SourceState; enabled: boolean; requires_key: boolean;
  last_attempt_at: string | null; last_success_at: string | null; newest_content_at: string | null;
  next_due_at: string | null; record_count: number; error: string | null; poll_interval_seconds: number;
  coverage: string | Record<string, unknown>; license_name: string; license_url: string | null; attribution: string;
}
export interface EventSummary {
  id: string; kind: string; category: Category; title: string; description: string;
  occurred_start: string | null; occurred_end: string | null; issued_at: string | null;
  source_updated_at: string | null; first_seen_at: string; last_seen_at: string; last_changed_at: string | null;
  valid_from: string | null; valid_to: string | null; countries: string[]; geometry: Geometry | null;
  location_precision: string; time_precision: string; severity: number; severity_label: string;
  severity_reason: string; original_severity: unknown; lifecycle_status: "active" | "expired" | "withdrawn" | "unknown";
  verification_status: string; anomaly_score: null; source_ids: string[]; source_count: number;
  independent_source_count: number; source_url: string | null; tags: string[]; change_type: string;
}
export interface Evidence {
  id: string; source_id: string; source_name: string; provider_record_id: string; source_url: string | null;
  retrieved_at: string; issued_at: string | null; source_updated_at: string | null;
  origins: string[]; payload_hash: string; raw: Record<string, unknown> | null; raw_retained?: boolean; source_snapshot_at?: string | null; attribution: string; license_url: string | null;
}
export interface EventDetail extends EventSummary {
  evidence: Evidence[];
  revisions: Array<{id: string; recorded_at: string; change_type: string; summary: string}>;
  relations: Array<{event_id: string; title: string; relation_type: string; reason: string; distance_km: number | null; time_delta_hours: number | null}>;
}
export interface EventQuery {
  window_hours: number; time_basis: TimeBasis; country?: string; region?: "europe"; category?: Category;
  severity_min: number; min_sources: number; lat?: number; lon?: number; radius_km?: number;
  include_inactive: boolean; limit: number; since?: string; until?: string;
}
export interface EventsResponse {
  items: EventSummary[]; total: number; shown: number; mapped: number; unlocated: number; truncated: boolean;
  query: EventQuery; source_health: SourceStatus[]; generated_at: string; limitations?: string[];
}
export interface SourcesResponse {items: SourceStatus[]; generated_at: string}
export interface Fact {text: string; event_id: string; source_urls: string[]}
export interface QueryResponse {
  supported: boolean; answer: string; interpretation: Record<string, unknown> | null;
  events: EventSummary[]; facts: Fact[]; inferences: string[]; limitations: string[];
  source_health: SourceStatus[]; generated_at: string; query_explanation?: string;
}
export interface BriefingResponse {
  id: string; answer: string; since: string; until: string; generated_at: string;
  sections: Array<{title: string; items: Array<{event_id: string; text: string}>}>;
  facts: Fact[]; inferences?: string[]; limitations: string[]; source_health: SourceStatus[];
  scope?: {window_hours: number; country?: string | null}; first_briefing?: boolean;
  total?: number; shown?: number; processed_count?: number | null; omitted_fact_count?: number | null; truncated?: boolean;
}
export interface HealthResponse {status: string; version: string; database: string; ai_mode: "off"; timezone: "Europe/Warsaw"}
